components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -0,0 +1,292 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/eth_abi_decoder.h"
+
+#include <limits>
+#include <map>
+
+#include "base/strings/string_util.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+
+// This file implements decoding rules of calldata according to the Contract
+// Application Binary Interface (ABI).
+//
+// Method arguments are packed in chunks of 32 bytes, with types smaller than
+// 32 bytes zero padded if necessary. Fixed-size types like uint256, address
+// are represented within the available 32 bytes, however, dynamic types like
+// bytes and arrays follow head-tail encoding. In this scheme, the data is
+// packaged at the tail-end of the transaction's calldata. The arguments are
+// references into the calldata where the content is.
+//
+// Parsing of complex cases such as composite types, nested arrays, etc., are
+// not implemented. The decoder only extracts the calldata tail reference in
+// such cases.
+//
+// Function selector (first 4 bytes) should NOT be part of the calldata being
+// parsed.
+//
+// References:
+//   https://docs.soliditylang.org/en/latest/abi-spec.html
+
+namespace brave_wallet {
+
+namespace {
+// GetAddressFromData extracts an Ethereum address from the calldata at the
+// specified index. The address type is static and 32-bytes wide, but we
+// only consider the last 20 bytes, discarding the leading 12 bytes of
+// 0-padded chars.
+//
+// In the future, addresses in Ethereum may become 32 bytes long:
+// https://ethereum-magicians.org/t/increasing-address-size-from-20-to-32-bytes
+//
+// The parsed address value is prefixed by "0x".
+bool GetAddressFromData(const std::string& input,
```

Comment:
Referring to my [earlier comment](https://github.com/brave/brave-core/pull/12797#discussion_r845773132), if we add checks to make the string operations are safe and the offset is within bounds, there'll be a case where we return `false`. So waiting for clarification on that before deciding this.

Reply:
Can't reply in referenced comment for some reason.
It makes sense to check range just before doing substr. Otherwise it is quite hard to track if proper checks were done somewhere before in callstack

Reply:
Fixed here 3dbb1b1a40eff8451f7f839f8033289d3de371e5.